packages/client/src/base/model/model-change-service.ts

```diff
@@ -0,0 +1,144 @@
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+import {
+    BoundsAwareViewportCommand,
+    Disposable,
+    DisposableCollection,
+    Emitter,
+    Event,
+    GModelRoot,
+    ICommand,
+    ICommandStack,
+    LazyInjector,
+    SetModelCommand,
+    SetViewportCommand,
+    TYPES,
+    UpdateModelCommand,
+    Viewport,
+    almostEquals,
+    isViewport
+} from '@eclipse-glsp/sprotty';
+import { inject, postConstruct, preDestroy } from 'inversify';
+
+/**
+ * Service that tracks changes to the model root and the viewport.
+ * Allows to register listeners that are notified when the model root or the viewport changes.
+ * The current model root can be queried at any time.
+ */
+export interface IModelChangeService {
+    /** The current model root */
+    readonly currentRoot: Readonly<GModelRoot>;
+    /**
+     * Event that is fired when the model root of the diagram changes i.e. after the `CommandStack` has processed a model update.
+     */
+    onModelRootChanged: Event<Readonly<GModelRoot>>;
+
+    /**
+     * Event that is fired when the viewport of the diagram changes i.e. after the `CommandStack` has processed a viewport update.
+     * By default, this event is only fired if the viewport was changed via a `SetViewportCommand` or `BoundsAwareViewportCommand`
+     */
+    onViewportChanged: Event<Readonly<Viewport>>;
+}
+
+export class ModelChangeService implements IModelChangeService, Disposable {
+    @inject(LazyInjector)
+    protected lazyInjector: LazyInjector;
+    protected _currentRoot: Readonly<GModelRoot>;
+    protected lastViewport?: Readonly<Viewport>;
+    protected toDispose = new DisposableCollection();
+
+    get currentRoot(): Readonly<GModelRoot> {
+        if (!this._currentRoot) {
+            throw new Error('Model root not available yet');
+        }
+        return this._currentRoot;
+    }
+
+    protected get commandStack(): ICommandStack {
+        return this.lazyInjector.get<ICommandStack>(TYPES.ICommandStack);
+    }
+
+    protected onModelRootChangedEmitter = new Emitter<Readonly<GModelRoot>>();
+    get onModelRootChanged(): Event<Readonly<GModelRoot>> {
+        return this.onModelRootChangedEmitter.event;
+    }
+
+    protected onViewportChangedEmitter = new Emitter<Readonly<Viewport>>();
+    get onViewportChanged(): Event<Readonly<Viewport>> {
+        return this.onViewportChangedEmitter.event;
+    }
+
+    @postConstruct()
+    protected initialize(): void {
+        this.toDispose.push(this.onModelRootChangedEmitter, this.onViewportChangedEmitter);
+        this.commandStack.onCommandExecuted(data => this.handleCommandExecution(data.command, data.newRoot), this);
```

Comment:
nit: I think we already have the correct `this` here without giving it as a parameter.